File: puzzle5/zoomlangVM/vm.py
class ZoomLangException(Exception):
  pass

class ZoomLangVM:
  def __init__(self, program, students=2):
    self.char_to_func = {'y': self.yes, 'f': self.faster, 's': self.slower, '+': self.up, '-': self.down, 'c': self.clap, 'b': self.take_break, 'a': self.away}
    self.program = list(program)
    self.students = students

    self.scores = [0] * students
    self.pc = 0
    self.pc_dir = 1
    self.step = 0
    self.unused_breaks = []
    self.used_breaks = []
    self.cur_reg = 0
    self.prev_reg = None
    self.reg_dir = 1
    self.steps_per_reg = [1] * students
    self.cur_reg_steps = 0

  def run(self):
    while self.pc < len(self.program):
      try:
        self.char_to_func[self.program[self.pc]]()
      except KeyError:
        if self.program[self.pc] == 'n':
          self.no()
        else:
          print(f'scores: {self.scores}')
          print(f"invalid symbol '{self.program[self.pc]}' at position {self.pc}")
        return
      except ZoomLangException as e:
        print(f'scores: {self.scores}')
        print(e)
        return
    print(f'scores: {self.scores}')

  # y
  def yes(self):
    if self.prev_reg is None:
      raise ZoomLangException("cannot execute yes because no previous register")
    self.scores[self.cur_reg] += self.scores[self.prev_reg]
    self._step()

  # n
  def no(self):
    print(f'scores: {self.scores}')
    print(f'pc: {self.pc}')
    print(f'pc_dir: {self.pc_dir}')
    print(f'step: {self.step}')
    print(f'unused_breaks: {self.unused_breaks}')
    print(f'used_breaks: {self.used_breaks}')
    print(f'cur_reg: {self.cur_reg}')
    print(f'prev_reg: {self.prev_reg}')
    print(f'reg_dir: {self.reg_dir}')
    print(f'steps_per_reg: {self.steps_per_reg}')
    print(f'cur_reg_steps: {self.cur_reg_steps}')

  # f
  def faster(self):
    if self.steps_per_reg[self.cur_reg] == 1:
      raise ZoomLangException(f"register {self.cur_reg} cannot go any faster")
    self.steps_per_reg[self.cur_reg] -= 1
    self._step()

  # s
  def slower(self):
    self.steps_per_reg[self.cur_reg] += 1
    self._step()

  # +
  def up(self):
    self.scores[self.cur_reg] += 1
    self._step()

  # -
  def down(self):
    self.scores[self.cur_reg] -= 1
    self._step()

  # c
  def clap(self):
    self.reg_dir *= -1
    self._step()

  # b
  def take_break(self):
    if self.pc not in self.used_breaks:
      self.unused_breaks.append(self.pc)
    self._step()

  # a
  def away(self):
    if self.unused_breaks:
      self.pc = self.unused_breaks.pop()
      self.used_breaks.insert(0, self.pc)
    self._step()

  # _step
  def _step(self):
    self.pc += self.pc_dir
    self.step += 1
    self.cur_reg_steps += 1
    if self.cur_reg_steps == self.steps_per_reg[self.cur_reg]:
      self.prev_reg = self.cur_reg
      self.cur_reg = (self.cur_reg + self.reg_dir) % self.students
      self.cur_reg_steps = 0

File: puzzle5/zoomlangVM/vm.hs
{-# LANGUAGE BlockArguments  #-}
{-# LANGUAGE NamedFieldPuns  #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE ViewPatterns    #-}

import           Control.Monad
import           Data.Foldable
import           Data.List
import qualified Data.Sequence        as S
import qualified Data.Vector.Storable as V
import           Numeric.Natural
import           System.Environment

data State = State
    { program       :: V.Vector Char
    , students      :: Int
    , scores        :: S.Seq Natural
    , pc            :: Int
    , pc_dir        :: Int
    , step          :: Natural
    , unused_breaks :: [Int]
    , used_breaks   :: [Int]
    , cur_reg       :: Int
    , prev_reg      :: Maybe Int
    , reg_dir       :: Int
    , steps_per_reg :: S.Seq Natural
    , cur_reg_steps :: Natural }

main = do
    [read -> students] <- getArgs
    forever do
        program <- getLine
        let (scores, err) = run State
                { program = V.fromList program
                , students
                , scores = S.replicate students 0
                , pc = 0
                , pc_dir = 1
                , step = 0
                , unused_breaks = []
                , used_breaks = []
                , cur_reg = 0
                , prev_reg = Nothing
                , reg_dir = 1
                , steps_per_reg = S.replicate students 1
                , cur_reg_steps = 0 }
            scoresOutput = "scores: " ++ show (toList scores)
        case err of
            Nothing -> putStrLn scoresOutput
            Just e -> putStr $ unlines ["\x1b[91m" ++ scoresOutput, e ++ "\x1b[0m"]

data ExecResult = Step State | Pass | Error String | Debug

run s@State {..} = case program V.!? pc of
    Nothing -> (scores, Nothing)
    Just c -> case exec c of
        Step s' -> run $ next $ tick s'
        Pass -> run $ tick s
        Error e -> (scores, Just e)
        Debug -> (scores, Just $ intercalate "\n"
            [ "pc = " ++ show pc
            , "pc_dir = " ++ show pc_dir
            , "step = " ++ show step
            , "unused_breaks = " ++ show (reverse $ toList unused_breaks)
            , "used_breaks = " ++ show (toList used_breaks)
            , "cur_reg = " ++ show cur_reg
            , "prev_reg = " ++ maybe "none" show prev_reg
            , "reg_dir = " ++ show reg_dir
            , "steps_per_reg = " ++ show (toList steps_per_reg)
            , "cur_reg_steps = " ++ show cur_reg_steps ])
  where exec 'y' = case prev_reg of
            Nothing -> Error "cannot execute yes because no previous register"
            Just pr -> Step s { scores = S.adjust' (+ scores `S.index` pr) cur_reg scores }
        exec 'n' = Debug
        exec 'f' = case steps_per_reg `S.index` cur_reg of
            1 -> Error $ "register " ++ show cur_reg ++ " cannot go any faster"
            _ -> adjustSpr pred
        exec 's' = adjustSpr succ
        exec '+' = adjustScores succ
        exec '-' = adjustScores pred
        exec 'c' = Step s { reg_dir = -reg_dir }
        exec 'b'
            | pc `elem` used_breaks = Pass
            | otherwise = Step s { unused_breaks = pc : unused_breaks }
        exec 'a' = Step case unused_breaks of
            [] -> s
            pc' : uub' -> s { pc = pc', unused_breaks = uub', used_breaks = pc' : used_breaks }
        exec c = Error $ "invalid symbol " ++ show c ++ " at position " ++ show pc
        adjustSpr f = Step s { steps_per_reg = S.adjust' f cur_reg steps_per_reg }
        adjustScores f = Step s { scores = S.adjust' f cur_reg scores }

tick s@State {..} = s { pc = pc + pc_dir }

next s@State {..}
    | crs' >= steps_per_reg `S.index` cur_reg = s' { prev_reg = Just cur_reg,
        cur_reg = (cur_reg + reg_dir) `mod` students, cur_reg_steps = 0 }
    | otherwise = s' { cur_reg_steps = crs' }
  where s' = s { step = succ step }
        crs' = succ cur_reg_steps
